test/test_recipes_integration.py

```diff
@@ -0,0 +1,46 @@
+import pytest
+import os
+import subprocess
+import json
+
+# Each test with recipe and appropriate parameters in one line
+# Using bracket annotation to set it optional (aka xfail)
+TEST_CASES = [
+    "activedata_usage",
+    ["backout_rate"],
+    ["code_coverage --path caps --rev 45715ece25fc"],
+    "code_coverage_by_suite --path caps --rev 45715ece25fc",
+    "config_durations",
+    "files_with_coverage",
+    ["intermittent_tests"],
+    ["intermittent_test_data"],
+    ["raw_coverage --path caps --rev 45715ece25fc"],
+    "test_durations",
+    ["tests_config_times -t test-windows10-64/opt-awsy-e10s"],
+    ["tests_in_duration"],
+    ["try_efficiency"],
+    ["try_usage"],
+    ["try_users"]
+]
+
+
+def load_tests(tests):
+    return [(test[0], True) if isinstance(test, list) else (test, False) for test in tests]
+
+
+@pytest.mark.parametrize("recipe, optional", load_tests(TEST_CASES),
+                         ids=[val[0] for val in load_tests(TEST_CASES)])
+@pytest.mark.skipif(os.getenv("TRAVIS_EVENT_TYPE") != "cron", reason="Not run by cron job")
+def test_recipe_integration(recipe, optional):
+    try:
+        command = "adr --format json {}".format(recipe).strip()
+        data = subprocess.check_output(command, shell=True, stderr=subprocess.STDOUT)
+        result = json.loads(data)
+        if not result or len(result) == 0:
+            raise ValueError("Invalid result {}".format(data))
+    except Exception as e:
+        # If this is optional test, will mark it xfail to make the test pass
+        if optional:
+            pytest.xfail(str(e))
```

Comment:
Clever idea! But pytest has a mechanism to do handle this already:
https://docs.pytest.org/en/latest/skipping.html#skip-xfail-with-parametrize

You can wrap your TEST_CASE lists with `pytest.param` to mark them as xfail directly. This will also let us define custom reason strings for each failing recipe if we want.

Reply:
Thanks. I updated the code. The current way is more concise but the previous way gives us the actual exception strings returned by the application, for example `No JSON object could be decoded` as you can see in https://travis-ci.com/TrangEmily/active-data-recipes/builds/88342554. Please let me know what do you prefer: dynamic (previous way) or hard-code reason strings or let them empty strings (like the current code change)?